rust/lance-index/src/vector/pq.rs

```diff
@@ -254,63 +253,103 @@ impl<T: ArrowFloatType + Cosine + Dot + L2> ProductQuantizerImpl<T> {
                 let distances = l2_distance_batch(sub_vec, subvec_centroids, sub_vector_length);
                 distance_table.extend(distances);
             });
+        Ok(distance_table)
+    }
+
+    /// Compute L2 distance from the query to all code.
+    ///
+    /// Type parameters
+    /// ---------------
+    /// - C: the tile size of code-book to run at once.
+    /// - V: the tile size of PQ code to run at once.
+    ///
+    /// Parameters
+    /// ----------
+    /// - distance_table: the pre-computed L2 distance table.
+    ///   It is a flatten array of [num_sub_vectors, num_centroids] f32.
+    /// - code: the PQ code to be used to compute the distances.
+    ///
+    /// Returns
+    /// -------
+    ///  The squared L2 distance.
+    #[inline]
+    fn compute_l2_distance<const C: usize, const V: usize>(
+        &self,
+        distance_table: &[f32],
+        code: &[u8],
+    ) -> Float32Array {
+        let num_centroids = num_centroids(self.num_bits);
+
+        let iter = code.chunks_exact(self.num_sub_vectors * V);
+        let distances = iter.clone().flat_map(|c| {
+            let mut sums = [0.0_f32; V];
+            for i in (0..self.num_sub_vectors).step_by(C) {
+                for (vec_idx, sum) in sums.iter_mut().enumerate() {
+                    let vec_start = vec_idx * self.num_sub_vectors;
+                    #[cfg(all(feature = "nightly", target_feature = "avx512f"))]
+                    {
+                        use std::arch::x86_64::*;
+                        if i + C <= self.num_sub_vectors {
+                            let mut offsets = [(i * num_centroids) as i32; C];
+                            for k in 0..C {
+                                offsets[k] += (k * num_centroids) as i32 + c[vec_start + k] as i32;
+                            }
+                            unsafe {
+                                let simd_offsets = _mm512_loadu_epi32(offsets.as_ptr());
+                                let v = _mm512_i32gather_ps(
+                                    simd_offsets,
+                                    distance_table.as_ptr() as *const u8,
+                                    4,
+                                );
+                                *sum += _mm512_reduce_add_ps(v);
+                            }
+                        } else {
+                            let mut s = 0.0;
+                            for k in 0..self.num_sub_vectors - i {
+                                *sum += distance_table
+                                    [(i + k) * num_centroids + c[vec_start + k] as usize];
+                            }
+                        }
+                    }
+                    #[cfg(not(any(feature = "nightly", target_feature = "avx512f")))]
```

Comment:
Does this mean that `avx512f` is now the only vectorized path?  In other words, if I had avx2 but not avx512f will I see a regression?

Reply:
avx2 uses auto-vectorization. But for avx512f, this `gather()` is more efficient than vectorization.